qiskit/transpiler/passes/synthesis/plugin.py

```diff
@@ -0,0 +1,362 @@
+plugins and expose them as valid options for the appropriate
+:func:`~qiskit.compiler.transpile` kwargs and pass constructors. If there are
+any installed plugins which can't be loaded/imported this will be logged to
+Python logging.
+
+To get the installed list of installed unitary synthesis plugins you can use the
+:func:`qiskit.transpiler.passes.synthesis.plugin.unitary_synthesis_plugin_names`
+function.
+
+Plugin API
+==========
+
+Unitary Synthesis Plugins
+-------------------------
+
+.. autosummary::
+   :toctree: ../stubs/
+
+   UnitarySynthesisPlugin
+   UnitarySynthesisPluginManager
+   unitary_synthesis_plugin_names
+
+"""
+
+import abc
+
+import stevedore
+
+
+class UnitarySynthesisPlugin(abc.ABC):
+    """Abstract unitary synthesis plugin class
+
+    This abstract class defines the interface for unitary synthesis plugins.
+    """
+
+    @property
+    @abc.abstractmethod
+    def max_qubits(self):
+        """Return the maximum number of qubits the unitary synthesis plugin supports.
+
+        If the size of the unitary to be synthesized exceeds this value the
+        ``default`` plugin will be used. If there is no upper bound return
+        ``None`` and all unitaries (``>= min_qubits`` if it's defined) will be
+        passed to this plugin when it's enabled.
+        """
+        pass
+
+    @property
+    @abc.abstractmethod
+    def min_qubits(self):
+        """Return the minimum number of qubits the unitary synthesis plugin supports.
+
+        If the size of the unitary to be synthesized is below this value the
+        ``default`` plugin will be used. If there is no lower bound return
+        ``None`` and all unitaries (``<= max_qubits`` if it's defined) will be
+        passed to this plugin when it's enabled.
+        """
+        pass
+
+    @property
+    @abc.abstractmethod
+    def supports_basis_gates(self):
+        """Return whether the plugin supports taking ``basis_gates``
+
+        If this returns ``True`` the plugin's ``run()`` method will be
+        passed a ``basis_gates`` kwarg with a list of gate names the target
+        backend supports. For example, ``['sx', 'x', 'cx', 'id', 'rz']``."""
+        pass
+
+    @property
+    @abc.abstractmethod
+    def supports_coupling_map(self):
+        """Return whether the plugin supports taking ``coupling_map``
+
+        If this returns ``True`` the plugin's ``run()`` method will receive
+        one kwarg ``coupling_map``. The ``coupling_map`` kwarg will be set to a
+        tuple with the first element being a
+        :class:`~qiskit.transpiler.CouplingMap` object representing the qubit
+        connectivity of the target backend, the second element will be a list
+        of integers that represent the qubit indices in the coupling map that
+        unitary is on. Note that if the target backend doesn't have a coupling
+        map set the ``coupling_map`` kwarg's value will be ``None``.
```

Comment:
Looking at the code, I think only the `CouplingMap` part of the tuple will be `None` if the backend doesn't support it, right?  We'll still be passing the qubit indices either way.

```suggestion
        unitary is on. Note that if the target backend doesn't have a coupling
        map set, the ``coupling_map`` kwarg's value will be ``(None, qubit_indices)``.
```

Reply:
Yeah, that's true good catch. I can change it to just pass None if you think that'd be better

Reply:
I feel like we should just provide the plugin with all the information we have to give them, and let them handle it.  Seems to be the design we've gone for elsewhere.  If that's the case, then passing the `(None, qubit_indices)` tuple is probably the right choice.